### DISCUSSION TITLE
Rooms dungeon model

### DISCUSSION THREAD

**Discussion by Inferdy** [mxgmn/MarkovJunior#38] [Show and tell]
A model with it's own flaws and beauty:

![6jz1hr](https://user-images.githubusercontent.com/24939153/174271415-74c51d65-5925-4c08-a9fa-2c349c360e86.gif)
![6jz1nk](https://user-images.githubusercontent.com/24939153/174271787-1c0c91b6-9ec9-4751-942b-a9ee773cf246.gif)
![6jz1ry](https://user-images.githubusercontent.com/24939153/174272061-9fd2d0c4-009d-4097-8029-8113f27a0081.gif)
![Rooms_1185080079](https://user-images.githubusercontent.com/24939153/174274962-311d2297-04aa-4ecc-ac03-1dff67ac7fbb.png)
![Rooms_1414903332](https://user-images.githubusercontent.com/24939153/174275004-24126c1c-e20d-49bc-aa54-53cb8789441e.png)
![Rooms_586017414](https://user-images.githubusercontent.com/24939153/174275361-d7aefc21-ed25-4dcf-a35b-db9223ec3efa.png)
![Rooms_1182520871](https://user-images.githubusercontent.com/24939153/174275388-d3f025f3-ae41-4e4f-8950-4d193968f051.png)

```xml
<model name="Rooms" size="55" steps="1000" gui="220" gif="true"/>
```

```xml
<sequence values="BRWGO" origin="Center">
   <markov>
      <sequence>
         <all in="BBB/BRB/BBB" out="WWW/W*W/WWW" steps="1"/>
         <one in="R***B" out="****G" steps="3"/>
         <prl in="R***G" out="**W**" steps="1"/>
         <prl in="R" out="W"/>
         <prl in="G" out="R"/>
      </sequence>
   </markov>
   <markov>
      <sequence>
         <one in="BBWBB/BWWWB/BWWWB/BWWWB/BBBBB" out="*****/*****/**R**/*****/*****" steps="1"/>
         <one in="R*B*W" out="G*W**" steps="1"/>
         <prl in="G*B*W" out="**W**" steps="1" p="0.35"/>
         <one in="G" out="W" steps="1"/>
         <one in="R" out="O" steps="1"/>
      </sequence>
   </markov>
   <one in="B*******B/****B****/****W****/*********/*BW***WB*/*********/****W****/****B****/B*******B" out="*********/****W****/*********/****W****/*W*WWW*W*/****W****/*********/****W****/*********"/>
</sequence>
```

**Comment by mxgmn** [comment]
Cool! Interesting trick with growing 3 rooms at a time. That's a mix between Growth and SAW -like models. I've added 3 similar models to the repository.

Some tips:
1. You don't need `steps="1"` in the first rulenode. In general, code with `steps="1"` can usually be reformulated without counters.
2. Second `steps="1"` node can be merged with the previous.
3. The second half of the model can be formulated without nested nodes.

I edited it a bit and made this dungeon generator:
```xml
<sequence values="BRWGO" origin="True">
  <markov>
    <sequence>
      <one in="R***B" out="*WWWG" steps="3"/>
      <all in="R" out="O"/>
      <all in="G" out="R"/>
    </sequence>
  </markov>
  <all in="BBB/BOB" out="***/*R*"/>
  <all in="RBBBR" out="OWWWO"/>
  <all in="R" out="O"/>
  <all in="BBBBB/BOWWW/BBBBB" out="*****/*BBBB/*****"/>
  <all in="OWWWO/W***W/W***W/W***W/OBBBO" out="*****/*****/*****/*****/*WWW*"/>
  <all in="*******/*OWWWO*/*W***W*/*W*B*W*/*W***W*/*OWWWO*/*******" out="WWWWWWW/W*****W/W*WWW*W/W*WGW*W/W*WWW*W/W*****W/WWWWWWW"/>
  <all in="O" out="W"/>
  <prl in="GWWWG/WWWWW/WWWWW/WWWWW/GWWWG" out="BBBBB/BBBBB/BBBBB/BBBBB/BBBBB"/>
  <all in="G" out="W"/>
</sequence>
```

![MultiHeadedDungeon](https://user-images.githubusercontent.com/14931730/174334153-b532300c-cb65-4ac7-9496-b56b9a6ceeb5.gif)

**Reply by Inferdy** [reply]
1. If I skip steps="1" then I'll get 2 equal frames in the output. With steps="1" I just clarify extra check is unneeded.
2. Could you give an example of merging such nodes?
> Second steps="1" node can be merged with the previous.
3. I'm not sure if it would work better or look better.  I would need to make dead ends marking rule the last rule to deal with 1 dead end at a time. This damages readabilty a little and it may be a performance issue aswell.

I like your variation of a dungeon generation. I had this idea of turning n-shapes into squares, too, but it feels better with corridors used in your model than room pieces used in mine. And, to be honest, I like this dungeon without final holes cutting inside rooms a bit more.

**Reply by mxgmn** [reply]
> Could you give an example of merging such nodes?

Like this:
```xml
<one in="R***B" out="**W*G" steps="3"/>
```

**Reply by Inferdy** [reply]
Haha! It's a bit different. In my case if one active (red) room builds a green pointer for the future room the other active room will have a connection with this future room, too. It was my strange wish 😜
